injected/src/features/message-bridge/create-page-world-bridge.js

```diff
@@ -0,0 +1,195 @@
+    DidInstall,
+    InstallProxy,
+    ProxyNotification,
+    ProxyRequest,
+    ProxyResponse,
+    SubscriptionRequest,
+    SubscriptionResponse,
+    SubscriptionUnsubscribe,
+} from './schema.js';
+import { isBeingFramed } from '../../utils.js';
+
+/**
+ * @import { MessagingInterface } from "./schema.js"
+ * @typedef {Pick<import("../../captured-globals.js"),
+ *    "dispatchEvent" | "addEventListener" | "removeEventListener" | "CustomEvent" | "String" | "Error" | "randomUUID">
+ * } Captured
+ */
+/** @type {Captured} */
+const captured = capturedGlobals;
+
+export const ERROR_MSG = 'Did not install Message Bridge';
+
+/**
+ * Try to create a message bridge.
+ *
+ * Note: This will throw an exception if the bridge cannot be established.
+ *
+ * @param {string} featureName
+ * @param {string} [token]
+ * @return {MessagingInterface}
+ * @throws {Error}
+ */
+export function createPageWorldBridge(featureName, token) {
+    /**
+     * This feature never operates without a featureName or token
+     */
+    if (typeof featureName !== 'string' || !token) {
+        throw new captured.Error(ERROR_MSG);
+    }
+    /**
+     * This feature never operates in a frame or insecure context
+     */
+    if (isBeingFramed() || !isSecureContext) {
+        throw new captured.Error(ERROR_MSG);
+    }
+
+    /**
+     * @param {string} eventName
+     * @return {`${string}-${string}`}
+     */
+    const appendToken = (eventName) => {
+        return `${eventName}-${token}`;
+    };
+
+    /**
+     * Create the sender to centralize the sending logic
+     * @param {{name: string} & Record<string, any>} incoming
+     */
+    const send = (incoming) => {
+        // when the token is absent, just silently fail
+        if (!token) return;
+        const event = new captured.CustomEvent(appendToken(incoming.name), { detail: incoming });
+        captured.dispatchEvent(event);
+    };
+
+    /**
+     * Events are synchronous (even across contexts), so we can figure out
+     * the result of installing the proxy before we return and give a
+     * better experience for consumers
+     */
+    let installed = false;
+    const id = random();
+    const evt = new InstallProxy({ featureName, id });
+    const evtName = appendToken(DidInstall.NAME + '-' + id);
+    const didInstall = (/** @type {CustomEvent<unknown>} */ e) => {
+        const result = DidInstall.create(e.detail);
+        if (result && result.id === id) {
+            installed = true;
+        }
+        captured.removeEventListener(evtName, didInstall);
+    };
+
+    captured.addEventListener(evtName, didInstall);
+    send(evt);
+
+    if (!installed) {
+        // leaving this as a generic message for now
+        throw new captured.Error(ERROR_MSG);
+    }
+
+    return createMessagingInterface(featureName, send, appendToken);
+}
+
+/**
+ * We are executing exclusively in secure contexts, so this should never fail
+ */
+function random() {
+    return captured.randomUUID?.() || 'n/a';
```

Comment:
Can we just make 'n/a' a never type or something. It seems weird to pacify the type system like this tbh. Optionally we could just throw as it should never happen as I understand it.

Reply:
yep will fix

Reply:
Fixed